src/SkeletalTrapezoidation.cpp

```diff
@@ -393,10 +510,71 @@ void SkeletalTrapezoidation::constructFromPolygons(const Polygons& polys)
         }
     }
 
-    vd_t vonoroi_diagram;
-    construct_voronoi(segments.begin(), segments.end(), &vonoroi_diagram);
+    vd_t voronoi_diagram;
+    construct_voronoi(segments.begin(), segments.end(), &voronoi_diagram);
 
-    for (vd_t::cell_type cell : vonoroi_diagram.cells())
+    // Try to detect cases when some Voronoi vertex is missing.
+    // When any Voronoi vertex is missing, rotate input polygon and try again.
+    constexpr double fix_angle = 0.5;  // Choose an off-kilter (non-Pi/non-Tau based) angle to counteract rotational symmetry.
+    const bool has_missing_voronoi_vertex = detectMissingVoronoiVertex(voronoi_diagram, points, segments);
+    const bool is_voronoi_planar = isVoronoiDiagramPlanarAngle(voronoi_diagram);
+    const bool is_voronoi_misconstructed = has_missing_voronoi_vertex || ! is_voronoi_planar;
+    std::unordered_map<Point, Point> vertex_mapping;  // NOTE: Should maybe add a functor to specify specialized hash as 3rd template parameter.
+    if (is_voronoi_misconstructed)
+    {
+        if (has_missing_voronoi_vertex)
+        {
+            spdlog::debug("Detected missing Voronoi vertex, input polygons will be rotated back and forth.");
+        }
+        if (! is_voronoi_planar)
+        {
+            spdlog::debug("Detected non-planar Voronoi diagram, input polygons will be rotated back and forth.");
+        }
+
+        Polygons polys_copy = polys;
```

Comment:
Thanks! Should be fixed by https://github.com/Ultimaker/CuraEngine/pull/1731/commits/d9b414f0b07ea270ec3dd450f9fa57b6d1862812

Reply:
Hoisting the declaration of `Polygons polys_copy` before the `if` block solves the segfault.
However the `isVoronoiDiagramPlanarAngle` checks fails unconditionally before and after rotation, even on the very simple model I uploaded above. Maybe false positives? It is probably related to your last commit.

Reply:
It might be. Or that method is just not the drop-in (counter-)clockwise detection replacement I thought it was.

Reply:
Yeah, I tried with and without the `Check for clockwiseness was inverted?`. It doesn't change the results. It probably requires a closer look at the generated cells.
I can't find the commits in their original context on github. Do you have a link?

Reply:
Link is here: https://github.com/prusa3d/PrusaSlicer/commit/77f5973c25ef0edc6db95de13c5976722b57a023#diff-bd3fe240334f0d6143dab5f96044b683bb6dd438f268318dcd566cc09a3c9416R68

(For context, maybe it's good to look at https://github.com/prusa3d/PrusaSlicer/commit/5db91f1944ff2def24a9e548b25024873f211f4c first though.)

I think I have a solution for the simple model at least, but I'm not fully sure about it: you can check that out with this branch `CURA-9178_temp_sidebranch`.

Reply:
I had to make a few changes to `CURA-9178_temp_sidebranch` make it compile
<details>

```diff
diff --git a/src/SkeletalTrapezoidation.cpp b/src/SkeletalTrapezoidation.cpp
index 0a1650054..29d995ade 100644
--- a/src/SkeletalTrapezoidation.cpp
+++ b/src/SkeletalTrapezoidation.cpp
@@ -449,7 +449,7 @@ bool SkeletalTrapezoidation::detectVoronoiErrors(const vd_t& voronoi_diagram, st
 
             } while (edge = edge->next(), edge != cell.incident_edge());
 
-            if ((count >= 2 && total_angle > 0.0 && !has_infinite)
+            if ((count >= 2 && total_angle > 0.0 && !has_infinite))
             {
                 spdlog::debug("Detected non-planar Voronoi diagram, input polygons will be rotated back and forth.");
                 return true;
@@ -466,9 +466,6 @@ bool SkeletalTrapezoidation::detectVoronoiErrors(const vd_t& voronoi_diagram, st
     return false;
 }
 
-{
-}
-
 void SkeletalTrapezoidation::constructFromPolygons(const Polygons& polys)
 {
     vd_edge_to_he_edge.clear();
```
</details>

This new check seems to helps in most of the cases I tried. I still got some false positive that can't be solved by rotation (I even did a retry loop with 0.5 rad increments)

I find the total_angle check a bit suspicious. Sometime it is a very small positive angle, probably from inaccuracies in `getAngleLeft` which returns a float. Sometime it is a negative angle somewhat close to -360°. -360° Should be good, right? But other times it deviates from that substantially, like -765°, -492°, etc...

Reply:
Yes, sorry about that. That's what I get for prettifying the code _after_ testing it.

I wonder if the deviation is because of the non-linear edges that occur when the diagram is between a line and a polygon-edge.

Reply:
I noticed than part of the deviations are caused by the rounding in `VoronoiUtils::p`, it is probably a better to do the calculation in doubles.

But even taking that into account, there are still what appears to be non convex cells when we look at two successive edges.  [`check_if_three_vectors_are_ccw`](https://github.com/prusa3d/PrusaSlicer/blob/ec2f533e3a355a2452bf6bd551da84020c69d8f8/src/libslic3r/Geometry/VoronoiUtilsCgal.cpp#L49-L68) takes 4 points over 3 edges for a reason, but I don't fully understand how it works yet...
I've been trying to replicate this function using `atan2(det, dot)` to compute the angle quadrant, but with little success.

I wonder if it would be better/simpler to implement better `edge->twin != nullptr` checks in the graph traversal, and on failure, wind back to a re-computation with rotated polygons.